File: src/memento/skill_import.py
from __future__ import annotations

import io
import shutil
import tempfile
import zipfile
from pathlib import Path

from memento.skill_packs import ValidatedSkillPack, validate_skill_pack


class SkillImportConflictError(FileExistsError):
    """Raised when a workspace already contains the recalled skill."""


def import_skill_pack(
    *,
    workspace: Path,
    skill_name: str,
    version: str,
    skill_md: str,
    zip_bytes: bytes,
) -> Path:
    """Validate and atomically import a recalled pack into `.pi/skills/<name>`."""
    validated = validate_skill_pack(
        skill_name=skill_name,
        version=version,
        skill_md=skill_md,
        zip_bytes=zip_bytes,
    )
    destination = workspace / ".pi" / "skills" / skill_name
    if destination.exists():
        raise SkillImportConflictError(f"skill destination already exists: {destination}")
    destination.parent.mkdir(parents=True, exist_ok=True)
    temp_dir = Path(tempfile.mkdtemp(prefix=f".{skill_name}-", dir=destination.parent))
    try:
        _extract_validated_pack(validated, temp_dir)
        temp_dir.rename(destination)
    except Exception:
        shutil.rmtree(temp_dir, ignore_errors=True)
        raise
    return destination


def _extract_validated_pack(pack: ValidatedSkillPack, destination: Path) -> None:
    allowed = {entry.path: entry for entry in pack.manifest.entries}
    with zipfile.ZipFile(io.BytesIO(pack.zip_bytes)) as archive:
        for info in archive.infolist():
            if info.is_dir():
                continue
            entry = allowed.get(info.filename)
            if entry is None:
                raise ValueError(f"validated manifest is missing ZIP entry: {info.filename}")
            target = destination / entry.path
            target.parent.mkdir(parents=True, exist_ok=True)
            data = archive.read(info)
            if len(data) != entry.size:
                raise ValueError(f"ZIP entry size changed during import: {entry.path}")
            target.write_bytes(data)
            target.chmod(0o644)
    for directory in sorted(
        (path for path in destination.rglob("*") if path.is_dir()), reverse=True
    ):
        directory.chmod(0o755)
    destination.chmod(0o755)

File: go/assets/skill_import.go
package assets

import (
	"archive/zip"
	"bytes"
	"compress/bzip2"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type SkillImportConflictError struct{ Destination string }

func (e *SkillImportConflictError) Error() string {
	return "skill destination already exists: " + e.Destination
}

type skillImportIO struct {
	evalSymlinks, abs func(string) (string, error)
	lstat             func(string) (os.FileInfo, error)
	mkdirAll          func(string, os.FileMode) error
	mkdirTemp         func(string, string) (string, error)
	removeAll         func(string) error
	rename            func(string, string) error
	writeFile         func(string, []byte, os.FileMode) error
	chmod             func(string, os.FileMode) error
	newArchive        func([]byte) (*zip.Reader, error)
	readArchive       func(*zip.File) ([]byte, error)
}

func defaultSkillImportIO() skillImportIO {
	return skillImportIO{filepath.EvalSymlinks, filepath.Abs, os.Lstat, os.MkdirAll, os.MkdirTemp, os.RemoveAll, os.Rename, os.WriteFile, os.Chmod, func(raw []byte) (*zip.Reader, error) {
		archive, err := zip.NewReader(bytes.NewReader(raw), int64(len(raw)))
		if err == nil {
			archive.RegisterDecompressor(12, skillBzipReader)
		}
		return archive, err
	}, readArchiveFile}
}
func skillBzipReader(r io.Reader) io.ReadCloser { return io.NopCloser(bzip2.NewReader(r)) }
func ImportSkillPack(workspace, skillName, version, skillMD string, zipBytes []byte) (string, error) {
	pack, err := ValidateSkillPack(skillName, version, skillMD, zipBytes)
	if err != nil {
		return "", err
	}
	return importValidatedSkillPack(workspace, pack, defaultSkillImportIO())
}
func importValidatedSkillPack(workspace string, pack ValidatedPack, ops skillImportIO) (string, error) {
	root, err := ops.evalSymlinks(workspace)
	if err != nil {
		return "", err
	}
	root, err = ops.abs(root)
	if err != nil {
		return "", err
	}
	parent := root
	for _, part := range []string{".pi", "skills"} {
		parent = filepath.Join(parent, part)
		info, lstatErr := ops.lstat(parent)
		if lstatErr == nil {
			if info.Mode()&os.ModeSymlink != 0 {
				return "", fmt.Errorf("skill import parent must not be a symlink: %s", parent)
			}
			if !info.IsDir() {
				return "", fmt.Errorf("skill import parent must be a directory: %s", parent)
			}
		} else if !errors.Is(lstatErr, os.ErrNotExist) {
			return "", lstatErr
		}
	}
	destination := filepath.Join(parent, pack.SkillName)
	if _, err = ops.lstat(destination); err == nil {
		return "", &SkillImportConflictError{destination}
	} else if !errors.Is(err, os.ErrNotExist) {
		return "", err
	}
	if err = ops.mkdirAll(parent, 0755); err != nil {
		return "", err
	}
	temp, err := ops.mkdirTemp(parent, "."+pack.SkillName+"-")
	if err != nil {
		return "", err
	}
	keep := false
	defer func() {
		if !keep {
			_ = ops.removeAll(temp)
		}
	}()
	if err = extractValidatedSkillPack(pack, temp, ops); err != nil {
		return "", err
	}
	if err = ops.rename(temp, destination); err != nil {
		if _, statErr := ops.lstat(destination); statErr == nil {
			return "", &SkillImportConflictError{destination}
		}
		return "", err
	}
	keep = true
	return destination, nil
}
func extractValidatedSkillPack(pack ValidatedPack, destination string, ops skillImportIO) error {
	allowed := map[string]ManifestEntry{}
	for _, entry := range pack.Manifest.Entries {
		allowed[entry.Path] = entry
	}
	archive, err := ops.newArchive(pack.ZIPBytes)
	if err != nil {
		return err
	}
	directories := map[string]bool{}
	for _, file := range archive.File {
		raw := file.Name
		raw, _, _ = strings.Cut(raw, "\x00")
		member, err := validateMemberPath(raw)
		if err != nil {
			return err
		}
		if strings.HasSuffix(raw, "/") {
			continue
		}
		entry, ok := allowed[member]
		if !ok {
			return fmt.Errorf("validated manifest is missing ZIP entry: %s", member)
		}
		data, err := ops.readArchive(file)
		if err != nil {
			return err
		}
		if uint64(len(data)) != entry.Size {
			return fmt.Errorf("ZIP entry size changed during import: %s", member)
		}
		target := filepath.Join(destination, filepath.FromSlash(member))
		if err = ops.mkdirAll(filepath.Dir(target), 0755); err != nil {
			return err
		}
		directories[filepath.Dir(target)] = true
		if err = ops.writeFile(target, data, 0644); err != nil {
			return err
		}
		if err = ops.chmod(target, 0644); err != nil {
			return err
		}
	}
	paths := make([]string, 0, len(directories))
	for path := range directories {
		paths = append(paths, path)
	}
	sort.Sort(sort.Reverse(sort.StringSlice(paths)))
	for _, path := range paths {
		if err = ops.chmod(path, 0755); err != nil {
			return err
		}
	}
	return ops.chmod(destination, 0755)
}
